### PR TITLE
Fleet-compatible generation for Grammar-Kit

### PR DESCRIPTION
**Pull request by KrylovBoris**
### What is this PR for?

This PR aims to introduce additional functionality to the Grammar-Kit plugin that allows users to generate a parser and adjacent code that is compatible with Fleet's API.

### Why were these changes made?

Currently, for plugin developers to port their IDEA plugins over to Fleet, they need to edit generated parser files or write them from scratch. To reduce the workload and allow for a more streamlined development process we introduce a way to generate Fleet-compatible code.

### What is Fleet-compatible code?

For the parser to be compatible with Fleet, it needs to import packages from Fleet's own parsing core and not have PSI-related features. This could be achieved by tweaking imports and stopping Grammar-Kit from generating PSI.

### What has been done?

For Fleet-specific generation logic, we created a separate `FleetParserGenerator`, a subclass of `ParserGenerator`. To avoid copying large chunks of code from the parent, we introduced a bunch of small methods that are responsible for adjusting original namespaces and generating small parts of the parser code. We also introduced `FleetGenOptions`, a subclass of GenOptions, to control file generation.

To interact with Fleet-specific parser, we added 2 actions to the plugin:

1. Generate Parser Code For Fleet;
2. Generate JFlex Lexer For Fleet; 

To generate lexers we use a new template file, which includes Fleet-specific imports.

We also introduced 2 Fleet-only attributes to BNF files:

1. `adjustPackagesForFleet` allows users to suppress moving generated files to "fleet/" directory. By default, when we generate files for Fleet, they are moved to "fleet/" directory, and "fleet." prefix is added to their respective package names. This is done to mimic Fleet's parser code layout and this option allows users to turn it off.
2. `generateFileTypeForFleet` was added by request of the Fleet team. Since almost all `IFileType` implementations in Fleet are the same, this attribute was added to reduce the amount of boilerplate code, plugin developers have to write to make their parsers compilable. This attribute holds all the parameters to generate `IFileType` implementation for Fleet.

Finally, we added an optional parameter [-f] to the main function to switch from `ParserGenerator` to `FleetParserGenerator`

### How have the other parts of the plugin been affected?

Since all Fleet-specific generation is stored in a separate class, it should not affect other parts of the program.

### What's your plans for the future? 

In the future we plan to add support of introduced functionality in grammar-kit-gradle-plugin.

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -7,9 +7,9 @@
 # https://www.jetbrains.com/intellij-repository/snapshots
 
 pluginName=Grammar-Kit
-pluginVersion=2022.3.2
-pluginSinceIdeaBuild=223
-ideaVersion=2022.3.3
+pluginVersion=2024.0.1
```
Comment (YannCebron) [inline gradle.properties:10]: Please do not modify plugin version

File: gradle.properties [hunk 2]
```diff
@@ -7,9 +7,9 @@
...
+pluginSinceIdeaBuild=241
```
Comment (YannCebron) [inline gradle.properties:11]: Why make it compatible with 241.* only?

Comment (KrylovBoris) [reply gradle.properties]: I misunderstood the purpose of this property. I reverted the changes, they are not needed for my changes to works

**Comment by gregsh** [comment]
Please avoid changing any configuration files, including libraries xmls and idea/*.xml

File: src/org/intellij/grammar/generator/ParserGenerator.java [hunk 3]
```diff
@@ -769,7 +779,7 @@ else if (javaType != Java.INTERFACE && i == 1) {
         sb.append(" implements ").append(shortener.shorten(aSuper));
       }
       else {
-        sb.append(", ").append(shortener.shorten(aSuper));
+         sb.append(", ").append(shortener.shorten(aSuper));
```
Comment (ice1000) [inline src/org/intellij/grammar/generator/ParserGenerator.java:782]: This is probably unintentional?

Comment (KrylovBoris) [reply src/org/intellij/grammar/generator/ParserGenerator.java]: Yes, it is. Thank you for pointing that out. I double-checked the indents in files, now there shouldn't be any stray whitespaces

File: src/org/intellij/grammar/Main.java [hunk 4]
```diff
@@ -27,7 +28,8 @@
 public class Main {
   public static void main(String[] args) {
     if (args.length < 2) {
-      System.out.println("Usage: Main <output-dir> <grammars or patterns>");
+      System.out.println(
+        "Usage: Main <output-dir> <grammars or pattern 1>[-f][--generateFileTypeElement --className=<fqn> --debugName=<debugName> --languageClass=<fqn>]... <grammars or pattern n>[-f][--generateFileTypeElement...]");
```
Comment (YannCebron) [inline src/org/intellij/grammar/Main.java:32]: `pattern 1`??

File: src/org/intellij/grammar/Main.java [hunk 5]
```diff
@@ -53,6 +61,66 @@ public static void main(String[] args) {
           return;
         }
 
+        while (i + 1 < args.length && (args[i + 1].startsWith("-f") || args[i + 1].startsWith("--generateFileTypeElement"))) {
```
Comment (YannCebron) [inline src/org/intellij/grammar/Main.java:64]: arg name `f` is not quite intuitive, can we make it `fleet`

File: src/org/intellij/grammar/LightPsi.java [hunk 6]
```diff
@@ -124,11 +124,12 @@ private static boolean shouldAddEntry(String path) {
     if (path.contains("/grammar-kit/")) return false;
 
     return path.contains("/out/classes/production/") ||
-           path.contains("idea.jar") ||
-           path.contains("platform-api.jar") ||
-           path.contains("platform-impl.jar") ||
            path.contains("util.jar") ||
-           path.contains("extensions.jar");
+           path.contains("util-8.jar") ||
```
Comment (YannCebron) [inline src/org/intellij/grammar/LightPsi.java:128]: why this change?

File: resources/META-INF/plugin-java.xml [hunk 7]
```diff
@@ -20,8 +20,11 @@
       <action id="grammar.Run.JFlex" class="org.intellij.grammar.actions.BnfRunJFlexAction"
               use-shortcut-of="grammar.Generate">
       </action>
+      <action id="grammar.fleet.Run.JFlex" class="org.intellij.grammar.actions.BnfRunFleetJFlexAction"/>
       <action id="grammar.Generate.ParserUtil" class="org.intellij.grammar.actions.BnfGenerateParserUtilAction"/>
       <action id="grammar.Generate.JFlexLexer" class="org.intellij.grammar.actions.BnfGenerateLexerAction"/>
+      <action id="grammar.fleet.Generate" class="org.intellij.grammar.actions.GenerateFleetAction"/>
```
Comment (YannCebron) [inline resources/META-INF/plugin-java.xml:26]: please add separator between IDEA/Fleet actions

File: resources/messages/GrammarKitBundle.properties [hunk 8]
```diff
@@ -14,6 +14,9 @@ action.grammar.Generate.JFlexLexer.text=Generate JFlex Lexer
 action.grammar.Generate.ParserUtil.text=Generate Parser Util
 action.grammar.Run.JFlex.text=Run JFlex Generator
 action.grammar.Generate.text=Generate Parser Code
+action.grammar.fleet.Generate.text=Generate Fleet-Compatible Parser Code
```
Comment (YannCebron) [inline resources/messages/GrammarKitBundle.properties:17]: IMHO: `Fleet: ....` name for actions would be quicker to scan

I'm concerned about "pollution" of the context menu, assuming that only few users will need Fleet-specific actions

File: src/org/intellij/grammar/LightPsi.java (hunk 6 above)
Comment (KrylovBoris) [reply src/org/intellij/grammar/LightPsi.java:128]: To reflect changes in the manifest file. Some libraries mentioned in this code no longer exist, and some required libraries were missing. So I rewrote that since I needed to be able to run grammar-kit code from a .jar file.

Comment (YannCebron) [reply src/org/intellij/grammar/LightPsi.java:128]: The JAR file names depend on the IDE version/distribution. I would expect it to no longer work with older IDE versions if removing entries.

File: resources/messages/GrammarKitBundle.properties (hunk 8 above)
Comment (KrylovBoris) [reply resources/messages/GrammarKitBundle.properties:17]: I suppose that we can hide Fleet-related actions behind a checkbox in Settings. But those who need Fleet-related would need to learn about this from the documentation.

Comment (YannCebron) [reply resources/messages/GrammarKitBundle.properties:17]: IMHO that would be overkill and too hard to discover.

Comment (KrylovBoris) [reply resources/messages/GrammarKitBundle.properties:17]: What if we show those actions if we detect fleet-related libraries referenced in the project?

Comment (YannCebron) [reply resources/messages/GrammarKitBundle.properties:17]: That would be the best solution probably.

File: src/org/intellij/grammar/LightPsi.java (hunk 6 above)
Comment (KrylovBoris) [reply src/org/intellij/grammar/LightPsi.java:128]: I assumed that JAR names depend on which SDK is used during build. I will re-add removed entries to the condition.

**Review by YannCebron — APPROVED** [review]

**Comment by YannCebron** [comment]
please add entry in CHANGELOG.md

File: src/org/intellij/grammar/generator/fleet/FleetParserGenerator.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                              @NotNull String outputPath,
+                              @NotNull String packagePrefix,
+                              Boolean generateIFileType,
+                              String classFileTypeName,
+                              String debugFileTypeName,
+                              String languageClass) {
+    super(psiFile, sourcePath, outputPath, packagePrefix);
+    myPossibleImports = new LinkedList<>();
+    var importList = psiFile.getAllPossibleAttributeValues(KnownAttribute.ELEMENT_TYPE_CLASS);
+    if (importList != null) myPossibleImports.addAll(psiFile.getAllPossibleAttributeValues(KnownAttribute.ELEMENT_TYPE_CLASS));
```
Comment (gregsh) [inline src/org/intellij/grammar/generator/fleet/FleetParserGenerator.java:50]: Call `getAllPossibleAttributeValues` for the second time?

File: src/org/intellij/grammar/generator/fleet/FleetParserGenerator.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.intellij.grammar.generator.ParserGeneratorUtil.getRootAttribute;
+import static org.intellij.grammar.generator.fleet.FleetConstants.FLEET_NAMESPACE;
+import static org.intellij.grammar.generator.fleet.FleetConstants.FLEET_NAMESPACE_PREFIX;
+
+public class FleetParserGenerator extends ParserGenerator {
+
+  private final Collection<String> myPossibleImports;
```
Comment (gregsh) [inline src/org/intellij/grammar/generator/fleet/FleetParserGenerator.java:25]: HashSet maybe?

File: src/org/intellij/grammar/generator/fleet/FleetParserGenerator.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+      try {
+        generateFileTypeClass();
+      }
+      finally {
+        closeOutput();
+      }
+    }
+  }
+
+  private void generateFileTypeClass() {
```
Comment (gregsh) [inline src/org/intellij/grammar/generator/fleet/FleetParserGenerator.java:156]: 1. Already outdated. `FleetPsiBuilder` is already gone.

2. Parser does not refer to this class, so it shall not be generated as part of parser generation. We can have a separate generator action for that if needed.

File: src/org/intellij/grammar/generator/ParserGenerator.java [hunk 12]
```diff
@@ -780,6 +790,16 @@ else if (javaType != Java.INTERFACE && i == 1) {
     myShortener = shortener;
   }
 
+  @NotNull
+  protected String generatePackageName(String className) {
+    return StringUtil.getPackageName(className);
+  }
+
+  @NotNull
+  protected String adjustName(String importName) {
```
Comment (gregsh) [inline src/org/intellij/grammar/generator/ParserGenerator.java:799]: `adjustName` must happen in BNF PSI, not in the generator.

As we need to support both "generate-time" and "edit-time" (in-IDE support)

File: src/org/intellij/grammar/actions/BnfGenerateLexerAction.java [hunk 13]
```diff
@@ -172,24 +179,33 @@ public void visitElement(@NotNull PsiElement element) {
...
     context.put("maxTokenLength", maxLen[0]);
 
     StringWriter out = new StringWriter();
-    InputStream stream = getClass().getResourceAsStream("/templates/lexer.flex.template");
+    InputStream stream = getClass().getResourceAsStream(getLexerFlexTemplate());
     ve.evaluate(context, out, "lexer.flex.template", new InputStreamReader(stream));
     return StringUtil.convertLineSeparators(out.toString());
   }
 
+  protected void putPackageName(@NotNull VelocityContext context, BnfFile bnfFile, @Nullable String packageName) {
```
Comment (gregsh) [inline src/org/intellij/grammar/actions/BnfGenerateLexerAction.java:200]: Adding a method for each property is not good. It's a famous OO-spaghetti code.


One method for context creation would be better.

**Comment by gregsh** [comment]
Overall

We don't want to have a separate `FleetParserGenerator`. We wish to parametrize `ParsersGenerator` and BNF PSI to simultaneously solve IntelliJ and Fleet tasks.

**Review by gregsh — CHANGES_REQUESTED** [review]

File: src/org/intellij/grammar/generator/ParserGenerator.java [hunk 14]
```diff
@@ -159,7 +159,7 @@ public ParserGenerator(@NotNull BnfFile psiFile,
     myOutputPath = outputPath;
     myPackagePrefix = packagePrefix;
 
-    G = new GenOptions(myFile);
+    G = createGenerator(myFile);
```
[the remark this answers is not in the document]
Comment (KrylovBoris) [reply src/org/intellij/grammar/generator/ParserGenerator.java]: I made createGenerator method obsolete, so I think we don't need to change the constructor

**Comment by gregsh** [comment]
I like that much more. And the `FleetBnfFileWrapper` trick is cool

Comments:
1. Some args miss nullability annos
2. `RuleInfo` belongs to `ParserGenerator`. Let's move it back
3. `ParserConstantSet` -> `IntelliJPlatformContants` ?
4. Let's have `NameShortener` back in `ParserGeneratorUtil`
5. `adjustPackagesForFleet` must not be publicly visible - let's encapsulate it more
6. I'd move all fleet-related classes in one flat package - `org.intellij.grammar.fleet`
7. `GenerateAction#getParserClass` seems unnecessary - we already adjust all BNF attributes
8. org.intellij.grammar.Main "Usage" string misses some spaces

**Comment by gregsh** [comment]
More Notes:
1. Let's  map files directly in `run -> doGenerate` instead of `BnfRunJFlexAction#getFileIterator` (and indent the code nearby 
2. Avoid copy-pasting `json.bnf` from IntelliJ repo
3. Inline `GeneratorBase.getFile()` and use `myFile` as before. The method is never overridden and must no be overridden
4. Reformat `org.intellij.grammar.BnfGeneratorAbstractTest#newTestGenerator` to avoid excessive indents and parens
5. Discuss `org.intellij.grammar.Main` arguments more

**Comment by zajac** [comment]
Any chance to get this merged already?

File: src/org/intellij/grammar/fleet/FleetBnfFileWrapper.java [hunk 15]
```diff
@@ -40,6 +41,10 @@ public class FleetBnfFileWrapper extends BnfFileImpl implements BnfFile {
 
   public FleetBnfFileWrapper(FileViewProvider viewProvider) {
     super(viewProvider);
+    if (viewProvider instanceof SingleRootFileViewProvider){
+      //feels super hacky, probably illegal
```
Comment (YannCebron) [inline src/org/intellij/grammar/fleet/FleetBnfFileWrapper.java:50]: please document why this is necessary

**Comment by gregsh** [comment]
One last change, please, and we are good to go!

1. Use the same naming: `GeneratorBase.file` -> `GeneratorBase.myFile` (lots of unnecessary changes in `ParserGenerator`)
